File: tests/fixtures/bespoke_gateway_loop/repo/app/loop.py
"""Ops assistant loop over the internal LLM gateway (no public host anywhere)."""

import json
import os

import requests

GATEWAY_URL = "https://gw.internal.example/llm/v1/chat"

SYSTEM_PROMPT = "You are the ops assistant for the payments desk."

TOOL_SCHEMAS = [
    {
        "type": "function",
        "function": {
            "name": "lookup_account",
            "description": "Read account master data.",
            "parameters": {"type": "object", "properties": {"account_id": {"type": "string"}}},
        },
    },
    {
        "type": "function",
        "function": {
            "name": "send_payment",
            "description": "Submit a payment instruction.",
            "parameters": {
                "type": "object",
                "properties": {
                    "account_id": {"type": "string"},
                    "amount": {"type": "string"},
                },
            },
        },
    },
]


def lookup_account(account_id: str) -> str:
    """Read account master data."""
    resp = requests.get(f"https://accounts.internal.example/v2/{account_id}")
    return resp.text


def send_payment(account_id: str, amount: str) -> str:
    """Submit a payment instruction to the payments core."""
    resp = requests.post(
        "https://payments-core.internal/api/payments",
        headers={"Authorization": "Bearer " + os.environ["PAYMENTS_TOKEN"]},
        json={"account": account_id, "amount": amount},
    )
    return resp.text


TOOLS = {"lookup_account": lookup_account, "send_payment": send_payment}


def run_agent(request_text: str) -> str:
    messages = [
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "user", "content": request_text},
    ]
    turns = 0
    while turns < 8:
        resp = requests.post(
            GATEWAY_URL,
            headers={"Authorization": "Bearer " + os.environ["GATEWAY_TOKEN"]},
            json={
                "model": "internal-x1",
                "messages": messages,
                "temperature": 0,
                "tools": TOOL_SCHEMAS,
            },
        )
        body = resp.json()
        message = body["choices"][0]["message"]
        calls = message.get("tool_calls")
        if not calls:
            return message["content"]
        messages.append(message)
        for call in calls:
            name = call["function"]["name"]
            handler = TOOLS[name]
            output = handler(**json.loads(call["function"]["arguments"]))
            messages.append(
                {"role": "tool", "tool_call_id": call["id"], "content": output}
            )
        turns += 1
    return "turn budget exhausted"

File: tests/fixtures/ts_bespoke_gateway_loop/repo/src/loop.ts
// Headline TS case: a hand-rolled agent loop over an internal LLM gateway.
// No SDK, no known host — detected by request shape, model attributed.
import { lookupAccount } from "./tools/lookup";
import { sendPayment } from "./tools/pay";

const TOOLS: Record<string, (args: unknown) => unknown> = {
  lookup_account: lookupAccount,
  send_payment: sendPayment,
};

const SYSTEM_PROMPT = "You are the ops assistant. Help with account operations.";

export async function runAgent(question: string): Promise<string> {
  const messages: Array<Record<string, unknown>> = [
    { role: "system", content: SYSTEM_PROMPT },
    { role: "user", content: question },
  ];
  while (true) {
    const resp = await fetch("https://gw.internal.example/llm/v1/chat", {
      method: "POST",
      headers: { Authorization: "Bearer " + process.env.GATEWAY_TOKEN },
      body: JSON.stringify({
        model: "internal-x1",
        messages,
        tools: Object.keys(TOOLS),
      }),
    });
    const data = await resp.json();
    const choice = data.choices[0];
    if (choice.finish_reason !== "tool_calls") {
      return choice.message.content;
    }
    messages.push(choice.message);
    for (const call of choice.message.tool_calls) {
      const fn = TOOLS[call.function.name];
      messages.push({ role: "tool", content: fn(call.function.arguments) });
    }
  }
}
